Linting, Testing, Coverage, and CI

ChrisAlderson opened the pull request:
Not sure if you are happy with this one because it breaks about everything for all the implementing packages.

 - Linting with [`eslint`](https://github.com/standard/eslint-config-standard) with a configuration based on [StandardJS](https://github.com/standard/standard)
 - Moved the `base.js` tests from `tape` to [`chai`](https://github.com/chaijs/chai) and [`mocha`](https://github.com/mochajs/mocha)
 - Added `mocha` tests for the `butter-provider` package
 - Added code coverage with [`nyc`](https://github.com/istanbuljs/nyc)
 - Added CI with `travis`

The tests defined in `basic.js` can still be executed by the implementations of the `butter-provider`, but since the directory holding the tests files has been changed from `tests` to `test` you probably need to make changes in the sub packages. The change of directories is because of mocha.

xaiki commented:
do we need to move the test directory ?
it's a bit of a hassle.

xaiki commented on index.js:
```diff
@@ -58,7 +59,7 @@ class Provider {
     const { argTypes, defaults } = this.config
     Object.keys(argTypes).map(k => {
-      if (!argTypes || !argTypes[k]) {
+      if (!parsed || !parsed[k]) {
```
this is needed anyway, can you PR it out ?

ChrisAlderson replied:
Well here you want to check if all the `args` in the `config` are parsed, right? So what should be done is:
 - parse the `argString` to an object
 - merge the default args from `config` with the parsed `argString`
 - loop through the needed arguments
    - check if a needed argument exists
       - if it does 👍 
       - if it does not 👎

xaiki commented:
it looks very good.
i'm ok to make one provider overhaul, but i'd like it that we don't have to change all the provider yet another time. can we find a test entrypoint that'd be stable so that we *never* have to change all the providers again ?

ChrisAlderson commented:
Looks like the test directory didn't need to be changed after all, so I changed it back. 
Also reverted something from my previous PR, where I changed `config.args` to `config.argTypes`

xaiki commented on index.js:
```diff
@@ -16,7 +16,7 @@ const defaultArgs = {
 
 const defaultConfig = {
-  argTypes: {},
+  args: {},
```
this patch needs to be dropped

xaiki commented:
6482bf5 needs to be dropped.

xaiki commented on index.js:
```diff
@@ -57,13 +58,15 @@ class Provider {
       : undefined
 
     const { args, defaults } = this.config
```
argTypes

ChrisAlderson replied:
fixed

xaiki commented on index.js:
```diff
@@ -57,13 +58,15 @@ class Provider {
       : undefined
 
     const { args, defaults } = this.config
+    const argsObject = Object.assign({}, defaults, parsed)
```
this can become args

ChrisAlderson replied:
fixed

xaiki commented on index.js:
```diff
@@ -74,8 +77,9 @@ class Provider {
+        const type = this.config.args[key]
 
-        args[key] = this._parseArgForType(key, value)
+        args[key] = this._parseArgForType(type, value)
```
this needs to get rebased on the fix-es6 branch

xaiki commented on package-lock.json:
```diff
@@ -0,0 +1,4069 @@
+{
```
is this supposed to be in ?

xaiki commented on package.json:
```diff
@@ -4,7 +4,9 @@
-    "test": "echo \"Error: no test specified\" && exit 1"
+    "coveralls": "cat ./coverage/lcov.info | node ./node_modules/coveralls/bin/coveralls",
+    "lint": "./node_modules/.bin/eslint tests/** index.js",
+    "test": "./node_modules/.bin/nyc --reporter=lcov --reporter=text ./node_modules/.bin/mocha tests/*.spec.*"
```
shouldn't this be `test/` ?

ChrisAlderson replied:
No I changed the directory back in `fbdc150`.

xaiki commented:
need to remove the lock file commited in 331e75540891d0d285515e410e777f9adeb6030c, if you're unconfortable doing that i can merge and rewrite history later.
